File: Zeze/Transaction/Collections/CollMap1.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Zeze.Serialize;

namespace Zeze.Transaction.Collections
{
	public class CollMap1<K, V> : CollMap<K, V>
	{
        public override V this[K key]
		{
			get => Map[key];
			set
			{
				if (value == null)
					throw new ArgumentNullException();

				if (this.IsManaged)
				{
					var txn = Transaction.Current;
					txn.VerifyRecordAccessed(this);
					var log = (LogMap1<K, V>)txn.LogGetOrAdd(Parent.ObjectId + VariableId, CreateLogBean);
					log.SetItem(key, value);
				}
				else
				{
					_map = _map.SetItem(key, value);
				}

			}
		}

        public override void Add(K key, V value)
		{
			if (key == null)
				throw new ArgumentNullException();
			if (value == null)
				throw new ArgumentNullException(); 
			
			if (IsManaged)
			{
				var txn = Transaction.Current;
				txn.VerifyRecordAccessed(this);
				var maplog = (LogMap1<K, V>)txn.LogGetOrAdd(Parent.ObjectId + VariableId, CreateLogBean);
				maplog.Add(key, value);
			}
			else
			{
				_map = _map.Add(key, value);
			}
		}

		public override void SetItem(K key, V value)
		{
			if (key == null)
				throw new ArgumentNullException();
			if (value == null)
				throw new ArgumentNullException(); 
			
			if (IsManaged)
            {
				var txn = Transaction.Current;
				txn.VerifyRecordAccessed(this);
				var maplog = (LogMap1<K, V>)txn.LogGetOrAdd(Parent.ObjectId + VariableId, CreateLogBean);
				maplog.SetItem(key, value);
			}
			else
            {
				_map = _map.SetItem(key, value);
            }
		}

		public override bool Remove(K key)
		{
			if (IsManaged)
			{
				var txn = Transaction.Current;
				txn.VerifyRecordAccessed(this); 
				var maplog = (LogMap1<K, V>)txn.LogGetOrAdd(Parent.ObjectId + VariableId, CreateLogBean);
				return maplog.Remove(key);
			}
			else
			{
				var newmap = _map.Remove(key);
				if (newmap == _map)
					return false;
				_map = newmap;
				return true;
			}
		}

		public override void Clear()
        {
			if (IsManaged)
			{
				var txn = Transaction.Current;
				txn.VerifyRecordAccessed(this);
				var maplog = (LogMap1<K, V>)txn.LogGetOrAdd(Parent.ObjectId + VariableId, CreateLogBean);
				maplog.Clear();
			}
			else
			{
				_map = _map.Clear();
			}
		}

        public override void FollowerApply(Log _log)
        {
			var log = (LogMap1<K, V>)_log;
			var tmp = _map;
			tmp = tmp.SetItems(log.Putted);
			tmp = tmp.RemoveRange(log.Removed);
			_map = tmp;
		}

		public override LogBean CreateLogBean()
		{
			var log = new LogMap1<K, V>();
			log.Belong = Parent;
			log.This = this;
			log.VariableId = VariableId;
			log.Value = _map;
			return log;
		}

		protected override void InitChildrenRootInfo(Record.RootInfo root)
		{ 
		}

		public override Bean CopyBean()
		{
			var copy = new CollMap1<K, V>();
			copy._map = _map;
			return copy;
		}

        public override void Add(KeyValuePair<K, V> item)
        {
			if (item.Key == null)
				throw new ArgumentNullException();
			if (item.Value == null)
				throw new ArgumentNullException();

			if (this.IsManaged)
			{
				var txn = Transaction.Current;
				txn.VerifyRecordAccessed(this);
				var log = (LogMap1<K, V>)txn.LogGetOrAdd(Parent.ObjectId + VariableId, CreateLogBean);
				log.Add(item.Key, item.Value);
			}
			else
			{
				_map = _map.Add(item.Key, item.Value);
			}
		}

		public override void AddRange(IEnumerable<KeyValuePair<K, V>> pairs)
        {
			foreach (var p in pairs)
			{
				if (p.Key == null)
					throw new ArgumentNullException();
				if (p.Value == null)
					throw new ArgumentNullException();
			}

			if (this.IsManaged)
			{
				var txn = Transaction.Current;
				txn.VerifyRecordAccessed(this);
				var log = (LogMap1<K, V>)txn.LogGetOrAdd(Parent.ObjectId + VariableId, CreateLogBean);
				log.AddRange(pairs);
			}
			else
			{
				_map = _map.AddRange(pairs);
			}
		}

		public override void SetItems(IEnumerable<KeyValuePair<K, V>> items)
        {
			foreach (var p in items)
			{
				if (p.Key == null)
					throw new ArgumentNullException();
				if (p.Value == null)
					throw new ArgumentNullException();
			}

			if (this.IsManaged)
			{
				var txn = Transaction.Current;
				txn.VerifyRecordAccessed(this);
				var log = (LogMap1<K, V>)txn.LogGetOrAdd(Parent.ObjectId + VariableId, CreateLogBean);
				log.SetItems(items);
			}
			else
			{
				_map = _map.SetItems(items);
			}
		}

		public override bool Remove(KeyValuePair<K, V> item)
        {
			if (this.IsManaged)
			{
				var txn = Transaction.Current;
				txn.VerifyRecordAccessed(this);
				var log = (LogMap1<K, V>)txn.LogGetOrAdd(Parent.ObjectId + VariableId, CreateLogBean);
				return log.Remove(item);
			}
			else
			{
				// equals处有box
				if (_map.TryGetValue(item.Key, out var oldv) && oldv.Equals(item.Value))
				{
					_map = _map.Remove(item.Key);
					return true;
				}
				else
				{
					return false;
				}
			}
		}
	}
}

File: ZezeJava/ZezeJava/src/main/java/Zeze/Transaction/Collections/CollMap1.java
package Zeze.Transaction.Collections;

import Zeze.Serialize.ByteBuffer;
import Zeze.Serialize.SerializeHelper;
import Zeze.Transaction.Bean;
import Zeze.Transaction.Log;
import Zeze.Transaction.Record;
import Zeze.Transaction.Transaction;
import Zeze.Util.Reflect;

public class CollMap1<K, V> extends CollMap<K, V> {
	protected final SerializeHelper.CodecFuncs<K> keyCodecFuncs;
	protected final SerializeHelper.CodecFuncs<V> valueCodecFuncs;
	private final int logTypeId;

	public CollMap1(Class<K> keyClass, Class<V> valueClass) {
		keyCodecFuncs = SerializeHelper.createCodec(keyClass);
		valueCodecFuncs = SerializeHelper.createCodec(valueClass);
		logTypeId = Zeze.Transaction.Bean.Hash32("Zeze.Raft.RocksRaft.LogMap1<"
				+ Reflect.GetStableName(keyClass) + ", " + Reflect.GetStableName(valueClass) + '>');
	}

	private CollMap1(int logTypeId, SerializeHelper.CodecFuncs<K> keyCodecFuncs,
					 SerializeHelper.CodecFuncs<V> valueCodecFuncs) {
		this.keyCodecFuncs = keyCodecFuncs;
		this.valueCodecFuncs = valueCodecFuncs;
		this.logTypeId = logTypeId;
	}

	@Override
	public void add(K key, V value) {
		if (isManaged()) {
			@SuppressWarnings("unchecked")
			var mapLog = (LogMap1<K, V>)Transaction.getCurrent().LogGetOrAdd(
					getParent().getObjectId() + getVariableId(), this::CreateLogBean);
			mapLog.Add(key, value);
		} else
			_map = _map.plus(key, value);
	}

	@Override
	public void put(K key, V value) {
		if (isManaged()) {
			@SuppressWarnings("unchecked")
			var mapLog = (LogMap1<K, V>)Transaction.getCurrent().LogGetOrAdd(
					getParent().getObjectId() + getVariableId(), this::CreateLogBean);
			mapLog.Put(key, value);
		} else
			_map = _map.plus(key, value);
	}

	@Override
	public void remove(K key) {
		if (isManaged()) {
			@SuppressWarnings("unchecked")
			var mapLog = (LogMap1<K, V>)Transaction.getCurrent().LogGetOrAdd(
					getParent().getObjectId() + getVariableId(), this::CreateLogBean);
			mapLog.Remove(key);
		} else
			_map = _map.minus(key);
	}

	@Override
	public void clear() {
		if (isManaged()) {
			@SuppressWarnings("unchecked")
			var mapLog = (LogMap1<K, V>)Transaction.getCurrent().LogGetOrAdd(
					getParent().getObjectId() + getVariableId(), this::CreateLogBean);
			mapLog.Clear();
		} else
			_map = org.pcollections.Empty.map();
	}

	@Override
	public void FollowerApply(Log _log) {
		@SuppressWarnings("unchecked")
		var log = (LogMap1<K, V>)_log;
		_map = _map.plusAll(log.getPutted()).minusAll(log.getRemoved());
	}

	@Override
	public LogBean CreateLogBean() {
		var log = new LogMap1<>(logTypeId, keyCodecFuncs, valueCodecFuncs);
		log.setBelong(getParent());
		log.setThis(this);
		log.setVariableId(getVariableId());
		log.setValue(_map);
		return log;
	}

	@Override
	protected void InitChildrenRootInfo(Record.RootInfo root) {
	}

	@Override
	public Bean CopyBean() {
		var copy = new CollMap1<>(logTypeId, keyCodecFuncs, valueCodecFuncs);
		copy._map = _map;
		return copy;
	}

	@Override
	public void Encode(ByteBuffer bb) {
		var tmp = getMap();
		bb.WriteUInt(tmp.size());
		var keyEncoder = keyCodecFuncs.encoder;
		var valueEncoder = valueCodecFuncs.encoder;
		for (var e : tmp.entrySet()) {
			keyEncoder.accept(bb, e.getKey());
			valueEncoder.accept(bb, e.getValue());
		}
	}

	@Override
	public void Decode(ByteBuffer bb) {
		clear();
		var keyDecoder = keyCodecFuncs.decoder;
		var valueDecoder = valueCodecFuncs.decoder;
		for (int i = bb.ReadUInt(); i > 0; i--) {
			var key = keyDecoder.apply(bb);
			var value = valueDecoder.apply(bb);
			put(key, value);
		}
	}
}
